hathor/indexes/rocksdb_tokens_index.py

```diff
@@ -0,0 +1,463 @@
+            assert len(key) == 32 + 1
+            return bytes(key)
+        assert len(utxo.tx_hash) == 32
+        key.extend(utxo.tx_hash)
+        key.append(utxo.output_index)
+        assert len(key) == 32 + 1 + 32 + 1
+        return bytes(key)
+
+    def _to_key_mint(self, token_uid: bytes, utxo: Optional[TokenUtxoInfo]) -> bytes:
+        return self._to_key_authority(token_uid, True, utxo)
+
+    def _to_key_melt(self, token_uid: bytes, utxo: Optional[TokenUtxoInfo]) -> bytes:
+        return self._to_key_authority(token_uid, False, utxo)
+
+    def _from_key_any(self, key: bytes) -> _KeyAny:
+        import struct
+        assert len(key) >= 1
+        tag = _Tag(key[0])
+        token_uid = bytes(key[1:33])
+        assert len(token_uid) == 32
+        if tag is _Tag.INFO:
+            assert len(key) == 32 + 1
+            return _KeyAny(token_uid, tag)
+        elif tag is _Tag.TXS:
+            assert len(key) == 32 + 1 + 4 + 32
+            timestamp: int
+            (timestamp,) = struct.unpack('>I', key[33:37])
+            tx_hash = key[37:]
+            assert len(tx_hash) == 32
+            return _KeyAny(token_uid, tag, timestamp=timestamp, tx_hash=tx_hash)
+        elif tag in {_Tag.MINT, _Tag.MELT}:
+            assert len(key) == 32 + 1 + 32 + 1
+            tx_hash = key[33:65]
+            index = key[65]
+            return _KeyAny(token_uid, tag, tx_hash=tx_hash, index=index)
+        else:
+            # XXX: if/elif is exhaustive for all possible tags and invalid tag value will fail sooner
+            raise NotImplementedError('unreachable')
+
+    def _to_value_info(self, info: _InfoDict) -> bytes:
+        return json_dumpb(info)
+
+    def _from_value_info(self, value: bytes) -> _InfoDict:
+        return cast(_InfoDict, json_loadb(value))
+
+    def _create_token_info(self, token_uid: bytes, name: str, symbol: str, total: int = 0) -> None:
+        key = self._to_key_info(token_uid)
+        old_value = self._db.get((self._cf, key))
+        assert old_value is None
+        value = self._to_value_info({
+            'name': name,
+            'symbol': symbol,
+            'total': total,
+        })
+        self._db.put((self._cf, key), value)
+
+    def _destroy_token(self, token_uid: bytes) -> None:
+        import rocksdb
+
+        # XXX: this can be a large operation, a writebatch works similar to a "SQL transaction" in that if there is a
+        #      failure it'll either persist all operations or none
+        key_info = self._to_key_info(token_uid)
+        key_mint = self._to_key_mint(token_uid, None)
+        key_melt = self._to_key_mint(token_uid, None)
+        key_txs = self._to_key_txs(token_uid, None)
+        batch = rocksdb.WriteBatch()
+        batch.delete((self._cf, key_info))
+        it = self._db.iterkeys(self._cf)
+        for seek_key in [key_mint, key_melt, key_txs]:
+            it.seek(seek_key)
+            for _, key in it:
+                if not key.startswith(key_mint):
```

Comment:
Hmm, you're right, this is a bug. It's odd that all tests pass.

Reply:
Fixed, but it still doesn't affect any tests.

Reply:
Should we try to add a test that fails the previous code?

Reply:
I added the test, which I would need a little more to fail, but would. What happens is that the "keys" that are supposedly removed on this step, were already removed before. So in essence they were being removed twice, and this method wouldn't have failed. But if I comment the other removals (which actually ran before), this method indeed fails before being fixed. There was also another bug from a typo, `key_melt` was the same as `key_mint`.

But also, I think the behavior of the memory storage when a removal happens is a bit unexpected, possibly enough to consider it a bug. This was discovered because of that test. So I made both behave how I think makes most sense.

Reply:
Just saw the typo, I missed that as well, good that you got it.

I didn't understand the memory storage bug, it's the change below, right?

![image](https://user-images.githubusercontent.com/3298774/148446712-a9e0cb45-fced-419f-838d-10964bf5a71e.png)

Is it just because accessing the dict after deleting it should raise `KeyError`? If that's the case, makes sense, but is there anything else?

Reply:
Yes. That's exactly it. Because the `del self._tokens[tx.hash]` was not last, it would, the `transactions = self._tokens[token_uid]._transactions` line would add it again because it is a `defaultdict`. So it would not raise a `KeyError`. I think it makes sense to raise a `KeyError`.